src/Swarm/Game/State.hs

```diff
@@ -354,3 +383,37 @@ emitMessage :: MonadState GameState m => Text -> m ()
 emitMessage msg = do
   q <- use messageQueue
   messageQueue %= (msg :) . (if length q >= maxMessageQueueSize then init else id)
+
+-- | The number of ticks elapsed since the game started.
+ticks :: Lens' GameState Integer
+
+-- | Takes a robot out of the activeRobots set and puts it in the waitingRobots
+--   queue.
+sleepUntil :: MonadState GameState m => Text -> Integer -> m ()
+sleepUntil rn time = do
+  internalActiveRobots %= S.delete rn
+  waitingRobots . at time . non [] %= (rn :)
+
+-- | Removes robots whose wake up time matches the current game ticks count
+--   from the waitingRobots queue and put them back in the activeRobots set.
+wakeUpRobotsDoneSleeping :: MonadState GameState m => m ()
+wakeUpRobotsDoneSleeping = do
+  time <- use ticks
+  mrns <- waitingRobots . at time <<.= Nothing
+  case mrns of
+    Nothing -> return ()
+    Just rns -> internalActiveRobots %= S.union (S.fromList rns)
+
+deleteRobot :: MonadState GameState m => Text -> m ()
+deleteRobot rn = do
+  mrobot <- robotMap . at rn <<.= Nothing
+  let robot = fromMaybe (error "expected robot to exist") mrobot
```

Comment:
As a general principle we should never, ever have calls to `error` in the codebase, even for things that "should never" happen: a bug in the code could then cause the game to crash, meaning a player would lose all their progress.  For this function in particular, I don't see any need for this anyway.  If the robot doesn't exist, can't we just do nothing?

Reply:
I suppose in that case you could argue that calling deleteRobot for a robot that doesn't exist is not a bug, so I'll do what you suggested. But my question still stands: what should we do when we discover that an invariant is broken at runtime?

Reply:
:+1: